cmds/tail/tail.go

```diff
@@ -0,0 +1,259 @@
+
+// Tail prints the lasts 10 lines of a file. Can additionally follow the
+// the end of the file as it grows.
+//
+// Synopsis:
+//     tail [-f] [-n lines_to_show] [FILE]
+//
+// Description:
+//     If no files are specified, read from stdin.
+//
+// Options:
+//     -f: follow the end of the file as it grows
+//     -n: specify the number of lines to show (default: 10)
+
+// Missing features:
+// - follow-mode (i.e. tail -f)
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"syscall"
+)
+
+var (
+	flagFollow   = flag.Bool("f", false, "follow the end of the file")
+	flagNumLines = flag.Int("n", 10, "specify the number of lines to show")
+)
+
+// TailConfig is a configuration object for the Tail function
+type TailConfig struct {
+	follow   bool // enable follow-mode (-f)
+	numLines int  // specifies the number of lines to print (-n)
+}
+
+// getBlockSize returns the number of bytes to read for each ReadAt call. This
+// helps minimize the number of syscalls to get the last N lines of the file.
+func getBlockSize(numLines int) int64 {
+	// This is currently computed as 81 * N, where N is the requested number of
+	// lines, and 81 is a relatively generous estimation of the average line
+	// length.
+	return 81 * int64(numLines)
+}
+
+// readLastLinesBackwards reads the last N lines from the provided file, reading
+// backwards from the end of the file. This is more efficient than reading from
+// the beginning, but can only be done on seekable files, (e.g. this won't work
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile *os.File, writer io.Writer, numLines int) error {
+	blkSize := getBlockSize(numLines)
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
+	readData := make([]byte, 0)
+	buf := make([]byte, blkSize)
+	pos := lastPos
+	foundLines := 0
+	// for each block, count how many new lines, until they add up to `numLines`
+	for {
+		if pos == 0 {
+			break
+		}
+		var thisChunkSize int64
+		if pos < blkSize {
+			thisChunkSize = pos
+		} else {
+			thisChunkSize = blkSize
+		}
+		pos -= thisChunkSize
+		n, err := inFile.ReadAt(buf, pos)
+		if err != nil {
+			if err == io.EOF {
+				// if we got EOF but we also received some data, go ahead and handle
+				// what we got
+				if n == 0 {
+					break
+				}
+			} else {
+				return err
+			}
+		}
+		// merge this block to what was read so far
+		readData = append(buf[:n], readData...)
+		// count how many lines we have so far, and stop reading if we have
+		// enough
+		foundLines += bytes.Count(buf[:n], []byte{'\n'})
+		if foundLines >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
```

Comment:
I preferred it a bit when this was still inside the loop. That way, no more than 2*chunkSize is ever in the buffer, rather than the entire file.

Reply:
Oh, nevermind, I'm sorry. I thought this was the other one.

Reply:
np, I just realized the same :)